krm-functions/lib/parser/parser.go

```diff
@@ -17,11 +17,218 @@ limitations under the License.
 package parser
 
 import (
+	"fmt"
+
 	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	// errors
+	errKubeObjectNotInitialized = "KubeObject not initialized"
 )
 
-type Parser interface {
-	// ParseKubeObject returns a fn sdk KubeObject; if something failed an error
-	// is returned
-	ParseKubeObject() (*fn.KubeObject, error)
+type Parser[T1 any] interface {
+	// GetKubeObject returns the present kubeObject
+	GetKubeObject() *fn.KubeObject
+	// GetGoStruct returns a go struct representing the present KRM resource
+	GetGoStruct() (T1, error)
+	// GetStringValue is a generic utility function that returns a string from
+	// a string slice representing the path in the yaml doc
+	GetStringValue(fields ...string) string
+	// GetIntValue is a generic utility function that returns a int from
+	// a string slice representing the path in the yaml doc
+	GetIntValue(fields ...string) int
+	// GetBoolValue is a generic utility function that returns a bool from
+	// a string slice representing the path in the yaml doc
+	GetBoolValue(fields ...string) bool
+	// GetStringMap is a generic utility function that returns a map[string]string from
+	// a string slice representing the path in the yaml doc
+	GetStringMap(fields ...string) map[string]string
+	// SetNestedString is a generic utility function that sets a string on
+	// a string slice representing the path in the yaml doc
+	SetNestedString(s string, fields ...string) error
+	// SetNestedInt is a generic utility function that sets a int on
+	// a string slice representing the path in the yaml doc
+	SetNestedInt(s int, fields ...string) error
+	// SetNestedBool is a generic utility function that sets a bool on
+	// a string slice representing the path in the yaml doc
+	SetNestedBool(s bool, fields ...string) error
+	// SetNestedMap is a generic utility function that sets a map[string]string on
+	// a string slice representing the path in the yaml doc
+	SetNestedMap(s map[string]string, fields ...string) error
+	// DeleteNestedField is a generic utility function that deletes
+	// a string slice representing the path from the yaml doc
+	DeleteNestedField(fields ...string) error
+}
+
+// NewFromKubeObject creates a new parser interface
+// It expects a *fn.KubeObject as input representing the serialized yaml file
+func NewFromKubeObject[T1 any](o *fn.KubeObject) Parser[T1] {
+	return &obj[T1]{
```

Comment:
May be a good idea to check  input "o" for nil and return error. Since this is a base method will be useful for all callers.